examples/linear_model/plot_lasso_and_elasticnet.py

```diff
@@ -1,97 +1,245 @@
-sparse signal corrupted with an additive noise. Estimated coefficients are
-compared with the ground-truth.
+The present example compares three l1-based regression models on a synthetic
+signal obtained from sparse and correlated features that are further corrupted
+with additive gaussian noise:
+
+ - a :ref:`lasso`;
+ - an :ref:`automatic_relevance_determination`;
+ - an :ref:`elastic_net`.
+
+It is known that the Lasso estimates turn to be close to the model selection
+estimates when the data dimensions grow, given that the irrelevant variables are
+not too correlated with the relevant ones. In the presence of correlated
+features, Lasso itself cannot select the correct sparsity pattern [1]_.
 
+Here we compare the performance of the three models in terms of the :math:`R^2`
+score, the fitting time and the sparsity of the estimated coefficients when
+compared with the ground-truth.
 """
 
+# Author: Arturo Amor <david-arturo.amor-quiroz@inria.fr>
+
 # %%
-# Data Generation
-# ---------------------------------------------------
+# Generate synthetic dataset
+# --------------------------
+#
+# We generate a dataset where the number of samples is lower than the total
+# number of features. This leads to a singular matrix that cannot be dealt with
+# an :ref:`ordinary_least_squares` as it requires some regularization.
+#
+# The target `y` is a linear combination with alternating signs of sinusoidal
+# signals. Only the 10 lowest out of the 100 frequencies in `X` are used to
+# generate `y`, while the rest of the features are not informative. This results
+# in a high dimentional sparse feature space, where some degree of
+# l1-penalization is necessary.
 
 import numpy as np
-import matplotlib.pyplot as plt
-
-from sklearn.metrics import r2_score
 
-np.random.seed(42)
+rng = np.random.RandomState(0)
+n_samples, n_features, n_informative = 50, 100, 10
+time_step = np.linspace(-2, 2, n_samples)
+freqs = 2 * np.pi * np.sort(rng.rand(n_features)) / 0.01
+X = np.zeros((n_samples, n_features))
 
-n_samples, n_features = 50, 100
-X = np.random.randn(n_samples, n_features)
+for i in range(n_features):
+    X[:, i] = np.sin(freqs[i] * time_step)
 
-# Decreasing coef w. alternated signs for visualization
 idx = np.arange(n_features)
-coef = (-1) ** idx * np.exp(-idx / 10)
-coef[10:] = 0  # sparsify coef
-y = np.dot(X, coef)
+true_coef = (-1) ** idx * np.exp(-idx / 10)
+true_coef[n_informative:] = 0  # sparsify coef
+y = np.dot(X, true_coef)
 
-# Add noise
-y += 0.01 * np.random.normal(size=n_samples)
+# %%
+# Some of the informative features have close frequencies to induce
+# (anti-)correlations.
 
-# Split data in train set and test set
-n_samples = X.shape[0]
-X_train, y_train = X[: n_samples // 2], y[: n_samples // 2]
-X_test, y_test = X[n_samples // 2 :], y[n_samples // 2 :]
+freqs[:n_informative]
 
 # %%
-# Lasso
-# ---------------------------------------------------
+# A random phase is introduced using `numpy.random.random_sample
```

Comment:
I think that we can generate the link with `:func:`numpy.random.random_sample` without the need to add the full hyperlink.